Add mojio component

simontb opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add a new component for [Mojio](https://moj.io/). In a first step this is a device tracker for cars with the Mojio device installed. Later additional properties from the car can be fetched.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml
device_tracker:
  - platform: mojio
    domain: <YOUR_PROVIDER> (e.g. "telekom-de" for [CarConnect](https://www.telekom.de/smarte-produkte/iot/carconnect)
    client_id: <YOUR_API_CLIENT_ID>
    client_secret: <YOUR_API_CLIENT_ID>
    username: <YOUR_EMAIL_ADDRESS>
    password: <YOUR_PASSWORD>
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This is my first PR to home-assistant (in fact the first PR to any Python project) and I'm not a Python developer. Any help would be greatly appreciated.
- I used FleetGo as a template, because it seemed quite similiar.
- ~~I wasn't able to test this locally, because I can't start hass on my PC.~~ _(works now, see post below)_
- ~~I couldn't test it on my hass Raspberry as well. Tried to copy-paste the code into custom components, but there was something wrong with the import of the SDK-dependency I developed. Code is on [GitHub](https://github.com/simontb/mojio-py) and deployed to [Pypi](https://pypi.org/project/mojio-sdk/).~~ _(irrelevant now)_

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

simontb committed:
Add mojio component

homeassistant commented:
Hi @simontb, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#37006).

Once you do that we will be able to review and accept this pull request.

Thanks!

simontb committed:
Ran isort on device_tracker.py

frenck commented:
Hi there @simontb!

Thanks for opening up a PR! That is always appreciated.

I understand starting developing an integration for the first time can be quite a big task. I think it is really required to set up a proper development environment before continuing on any of this. This is needed in order to be able to test and use the proper tooling to format and validate what you are doing.

We have instructions up for setting up a development environment (several methods) listed here:

<https://developers.home-assistant.io/docs/development_environment>

Many more documentation and instructions can be found on that website.

Feel free to join our Discord server is you have any question:

<https://www.home-assistant.io/join-chat>

simontb committed:
Fix mojio-sdk import

simontb committed:
First working version of Mojio integration

Update SDK dependency version
Track car every 5 minutes
Fix license plate as identifier

simontb commented:
I finally got the dev environment set-up and the (still very basic) integration is working:

![mojio_entity_status](https://user-images.githubusercontent.com/6491719/85922215-e6299c00-b881-11ea-892f-590152cb2cd1.png)

Commit:
Exclude mojio from code coverage and remove commented out code

simontb commented:
@frenck could you provide me some guidance how to get this PR ready to be merged?

GSzabados commented:
@simontb, I am looking forward for your integration. It is nice to see that someone is finally adding Mojio to HA. Keep up with the good work!

One question, do you have a list of domains?

simontb commented:
@GSzabados thanks! It's nice to hear that my work will be useful to others as well :)

No, unfortunately I don't have a list of domains. It seems mojio is a multi-tenant solution, so you have to obtain the domain and client credentials yourself. But maybe the community can come up with a list.

You can get these by either observing what your client (e.g. mobile app or website) sends, or by asking your provider.

GSzabados commented:
@simontb, thanks for the answer. Actually the last bit is what I was a bit afraid of.

Mojio seems to be selling its product in different tiers. It might effect API access as well. I have to check then. I saw this controversial setup from Mojio's IFTTT integration. (T-Mobile Poland hasn't got access to the IFTTT channel.) But the app is exactly the same as the German one. But of course the dongle works only with the country's app.

I will find out how things work when it will be released. I will definitely give you some feedback then. Thanks again for the good work!

simontb commented:
Could you send me a link to that discussion, please?

GSzabados commented:
@simontb, I contacted directly Mojio support. They indeed not support individual users. Mobile operators have no clue about the hardware/software neither, at least the support line.

A reply from Mojio about IFTTT not available for Poland:

Thank you for writing to Mojio.

We received your request through our information channel.

We are a Business-to-Business company and provide the technology to carrier. We do not directly support any of T-Mobile's customers such as yourself.

We do not directly support any of carrier's customers. We provide our software to our business partners but each provider has customized their product in accordance with their business needs. 

As such, the best way for you to obtain support on your product is to get in contact with the carrier's frontline support.


Need to find the right contact information for your Service Provider? Please refer to: https://www.moj.io/support/

Thanks again for using a Mojio powered device!

simontb commented:
Thanks for reaching out to them and clarifying this, @GSzabados 

Just to be sure: you're device is this one? https://www.t-mobile.pl/pl/oferta_specjalna/smart-car

Looks like the same I have from Deutsche Telekom (T-Mobile Germany). So in general it should work with this integration. I haven't tested it though.

You could still ask T-Mobile Poland for API access or obtain the credentials yourself in case this complies with their terms of service.

In the meantime could you do me a favor and post a link to the discussion on the Mojio integration of IFTTT?

timmo001 reviewed (changes requested):
@simontb Your PR is using the 'old' method of creating a component. The standard now is to create an integration. See the developer docs on getting started with this setup:

https://developers.home-assistant.io/docs/creating_integration_file_structure

You can use a scaffold to get you started:

https://developers.home-assistant.io/docs/creating_component_index/

From there, you can create a config flow:

https://developers.home-assistant.io/docs/config_entries_config_flow_handler

For this specific integration, it looks like you will be needing to use the oauth2 standard, although this is only from a glance at your code. There will likely need to be changes to the library to support this.

https://developers.home-assistant.io/docs/api_lib_index
https://developers.home-assistant.io/docs/api_lib_auth/#oauth2

There is a scaffold script for oauth2 also to get you started with this.

simontb commented:
@timmo001 thanks a lot for the review.

Could you provide me the names of some integrations that use this new method? I tried first with the links you provided, but since I'm no python developer I wasn't able to get this working. Then I decided to copy and adjust an existing integration that is very similiar (FleetGo) until it worked.

Maybe I'd be able to use the new method, but only if I find something to copy from.

timmo001 commented:
All integrations use the new method but oauth2 specifically, `almond` and `home_connect` are a couple of examples.

There is no need to copy the integrations, as the `script.scaffold` will create a template integration for you. You will then need to just fill out the details and specifics and add the platform once you have authentication working.

frenck commented:
This integration needs some work/rewriting before we can review/accept it again.

For now, I'm closing this PR for that reason. Please, do re-open a PR when ready with the changes 👍
And when in need of any help with that, please join or Discord chat server!

martin131 commented:
@simontb Any news about Mojio integration? Do you have any other solution please?

simontb commented:
Hi @martin131,

I don't have the capacity to address the required changes right now. The integration is working. Feel free to use it as a custom component.

I'd also be very happy if you or someone else would fork the code and rework the integration to fit the requirements stated in the previous comments.

martin131 commented:
@simontb Thanks for the answer. Do you have confirmed that the integration works with the czech version (T-Mobile Chytré auto)? I have not ordered it yet.

simontb commented:
@martin131 it should work, but I can't verify this, because I don't have a czech device. With German Telekom this works. Keep in mind, that you need to retrieve API credentials in order to use this integration.

GSzabados commented:
@simontb, How did you get the client ID and client Secret values? What was the way for the German Telekom?

simontb commented:
@GSzabados had a look at the web app :D

But I don't know, if this is still working

GSzabados commented:
@simontb, unfortunately it doesn't show it there obviously...

But by any chance is it id_token and access_token?

simontb commented:
@GSzabados no there were explicit client_id and client_secret in the POST request of the authentication.

GSzabados commented:
@simontb it seems like not anymore. Have you built against the v1 or v2 API?

I see, client_id, that's ok, but the authentication response type is id_token, and it gets a bearers token.

GSzabados commented:
@simontb 

It looks like that the website uses the id_token authentication method to get the bearer token, it is not using the "password" method as your mojio-py:

https://tmobile-pl-identity.moj.io/account/signinmojio?ReturnUrl=%2Fconnect%2Fauthorize%2Fcallback%3Fclient_id%3<YOUR_API_CLIENT_ID>%26redirect_uri%3Dhttps%253A%252F%252Ftmobile-pl-web.moj.io%252Flive-view%26response_type%3Dtoken%2520id_token%26scope%3Dopenid%2520full%26nonce%3Dmy-mojio

And unfortunately I cannot guess the client_secret.

GSzabados commented:
@simontb 

Never mind, I have found it. 

Here are some,

T-Mobile PL:

```
device_tracker:
  - platform: mojio
    domain: tmobile-pl
    client_id: a29871cb-27c6-4e36-8249-205bafe659b3
    client_secret: 8eb2d8b3-7baf-4c41-9373-877ed9960a5f
    username: <YOUR_EMAIL_ADDRESS>
    password: <YOUR_PASSWORD>
```

T-Mobile CZ:

```
device_tracker:
  - platform: mojio
    domain: tmobile-cz
    client_id: 01e950e3-5e50-4cc0-8b04-8b48c35d00cd
    client_secret: a1b7d477-9410-4e1d-b537-304e041fcb89
    username: <YOUR_EMAIL_ADDRESS>
    password: <YOUR_PASSWORD>
```

Telekom DE

```
device_tracker:
  - platform: mojio
    domain: telekom-de
    client_id: cae6e4b9-00ca-4019-be80-26ed36a30979
    client_secret: 4d37a73e-72a3-466f-8ed2-f0fd4026cb77
    username: <YOUR_EMAIL_ADDRESS>
    password: <YOUR_PASSWORD>
```

The `client_id` and `client_secret` are operator specific. It identifies the App, on the Mojio API. 

I've set up the integration as a custom component and it is working. It shows the location of the car.

GSzabados commented:
> Hi, Integration works for me now, but it is not reporting position after few days (or hours?). In the original mobile app, it is working well. Where I can see some logs, please?

I have experienced the same. I think it is 24 hours. I haven't had any time to look into the issue yet, but I opened one on @simontb's module's github repo.

I believe it must due to the token expires, but somehow it is not recognised as expired. But that is only an educated guess.

simontb commented:
Hi @martin131,

I think @GSzabados is right. Please follow the issue he mentioned.

simontb commented:
@GSzabados and @martin131 please update mojio-sdk to version 0.0.8 and let me know if this solved your issue.

Bojkas1985 commented:
Hi guys, any news?

GSzabados commented:
> Hi guys, any news?

Hi,

It hasn't been merged officially, but can be used as a custom component. I am using it since last April and it works well. If you need help how to set it up, the best is to post here:

https://community.home-assistant.io/t/moj-io-integration/89019

richpanda1214 commented:
> If you need help how to set it up, the best is to post here:

Can you explain how you added it as a custom component? I am new to Home Assistant and would appreciate the help.

GSzabados commented:
> Can you explain how you added it as a custom component?

Use this as a custom component:

https://github.com/simontb/mojio-home-assistant
